include/RequestLimiter.inc.php

```diff
@@ -1,119 +1,101 @@
-
+	local tokens_key = KEYS[1]
+	local timestamp_key = KEYS[2]
+	
+	local rate = tonumber(ARGV[1])
+	local capacity = tonumber(ARGV[2])
+	local now = tonumber(ARGV[3])
+	local requested = tonumber(ARGV[4])
+	
+	local fill_time = capacity/rate
+	local ttl = math.floor(fill_time*2)
+	
+	local last_tokens = tonumber(redis.call("get", tokens_key))
+	if last_tokens == nil then
+		last_tokens = capacity
+	end
+	
+	local last_refreshed = tonumber(redis.call("get", timestamp_key))
+	if last_refreshed == nil then
+		last_refreshed = 0
+	end
+	
+	local delta = math.max(0, now-last_refreshed)
+	local filled_tokens = math.min(capacity, last_tokens+(delta*rate))
+	local allowed = filled_tokens >= requested
+	local new_tokens = filled_tokens
+	if allowed then
+		new_tokens = filled_tokens - requested
+	end
+	
+	redis.call("setex", tokens_key, ttl, new_tokens)
+	redis.call("setex", timestamp_key, ttl, now)
+	
+	return { allowed, new_tokens }';
+	
+	// Lua script is from https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d
 	protected $concurrent_requests_limiter_lua = '
-    local key = KEYS[1]
-
-    local capacity = tonumber(ARGV[1])
-    local timestamp = tonumber(ARGV[2])
-    local id = ARGV[3]
-    
-    local count = redis.call("zcard", key)
-    local allowed = count < capacity
-    
-    if allowed then
-      redis.call("zadd", key, timestamp, id)
-    end
-    
-    return { allowed, count }';
-
+	local key = KEYS[1]
+	
+	local capacity = tonumber(ARGV[1])
+	local timestamp = tonumber(ARGV[2])
+	local id = ARGV[3]
+	
+	local count = redis.call("zcard", key)
+	local allowed = count < capacity
+	
+	if allowed then
+		redis.call("zadd", key, timestamp, id)
+	end
+	
+	return { allowed, count }';
+	
 	public function __construct() {
 		$this->redis = Z_Redis::get();
-		$this->request_rate_limiter_lua_sha1 = sha1($this->request_rate_limiter_lua); //we can hardcode sha1 to prevent calculating it every time
+		// SHA1 can be hardcoded to prevent calculating it every time
+		$this->request_rate_limiter_lua_sha1 = sha1($this->request_rate_limiter_lua);
 		$this->concurrent_requests_limiter_lua_sha1 = sha1($this->concurrent_requests_limiter_lua);
 	}
-
+	
 	/**
-	 * Limit request rate for a given bucket
-	 * @param $params - bucket (userid or key), limit (requests per second), burst (multiplier for 'limit'), warn (threshold for remaining requests warning)
-	 * @return array|null
+	 * Limit request rate for a given bucket.
+	 * Rate and capacity parameters allows to have flexible rate limits.
+	 *
+	 * rate - request accumulation rate per second (can be below 1)
+	 * capacity - how many requests it can accumulate
+	 *
+	 * capacity=10, rate=1 means 10 request burst with 1 request per second rate
+	 * capacity=100, rate=0.5 means 100 request burst with 1 request per two seconds rate
+	 *
+	 * @param $params - bucket, capacity, rate
+	 * @return bool|null - returns true if request is allowed
 	 */
 	public function limitRate($params) {
```

Comment:
Maybe something clearer to call this? It's not really doing limiting itself. `checkBucketRate`?